kythe/cxx/indexer/cxx/IndexerASTHooks.h

```diff
@@ -93,34 +99,73 @@ void InsertAnchorMarks(std::string& Text, std::vector<MiniAnchor>& Anchors);
+  ProfilingCallback ReportProfileEvent = [](const char*, ProfilingEvent) {};
+  /// \brief Whether to use the CompilationUnit VName corpus as the default
+  /// corpus.
+  bool UseCompilationCorpusAsDefault = false;
+  /// \brief Whether we should stop on missing cases or continue on.
+  BehaviorOnUnimplemented IgnoreUnimplemented = BehaviorOnUnimplemented::Abort;
+  /// \brief Whether we should visit template instantiations.
+  BehaviorOnTemplates TemplateMode = BehaviorOnTemplates::VisitInstantiations;
+  /// \brief Whether we should emit all data.
+  Verbosity Verbosity = kythe::Verbosity::Classic;
+  /// \brief Should we emit documentation for forward class decls in ObjC?
+  BehaviorOnFwdDeclComments ObjCFwdDocs = BehaviorOnFwdDeclComments::Emit;
+  /// \brief Should we emit documentation for forward decls in C++?
+  BehaviorOnFwdDeclComments CppFwdDocs = BehaviorOnFwdDeclComments::Emit;
+  /// \return true if we should stop indexing.
+  std::function<bool()> ShouldStopIndexing = [] { return false; };
+  /// \brief The number of (raw) bytes to use to represent a USR. If 0,
+  /// no USRs will be recorded.
+  int UsrByteSize = 0;
+  /// \brief Controls whether dataflow edges are emitted.
+  EmitDataflowEdges DataflowEdges = EmitDataflowEdges::No;
+  /// \brief Controls whether abs nodes are emitted.
+  UseAbsNodes AbsNodes = UseAbsNodes::Abs;
+  /// \brief if nonempty, the pattern to match a path against to see whether
+  /// it should be excluded from template instance indexing.
+  std::shared_ptr<re2::RE2> TemplateInstanceExcludePathPattern = nullptr;
+  /// \param Worklist A function that generates a new worklist for the given
+  /// visitor.
+  std::function<std::unique_ptr<IndexerWorklist>(IndexerASTVisitor*)>
+      CreateWorklist = [](IndexerASTVisitor* indexer) {
+        return IndexerWorklist::CreateDefaultWorklist(indexer);
+      };
+};
+
 /// \brief An AST visitor that extracts information for a translation unit and
 /// writes it to a `GraphObserver`.
 class IndexerASTVisitor : public RecursiveTypeVisitor<IndexerASTVisitor> {
   using Base = RecursiveTypeVisitor;
 
  public:
-  IndexerASTVisitor(clang::ASTContext& C, BehaviorOnUnimplemented B,
-                    BehaviorOnTemplates T, Verbosity V,
-                    BehaviorOnFwdDeclComments ObjC,
-                    BehaviorOnFwdDeclComments Cpp, const LibrarySupports& S,
-                    clang::Sema& Sema, std::function<bool()> ShouldStopIndexing,
-                    GraphObserver* GO = nullptr, int UsrByteSize = 0,
-                    EmitDataflowEdges EDE = EmitDataflowEdges::No,
-                    std::shared_ptr<re2::RE2> TIEPP = nullptr)
-      : IgnoreUnimplemented(B),
-        TemplateMode(T),
-        Verbosity(V),
-        ObjCFwdDocs(ObjC),
-        CppFwdDocs(Cpp),
-        Observer(GO ? *GO : NullObserver),
+  IndexerASTVisitor(clang::ASTContext& C, const LibrarySupports& S,
+                    clang::Sema& Sema, const IndexerOptions& IO,
```

Comment:
nit: pass-by-value and std::move into the member? Retained const-ref parameter are risky as they bind to temporaries.

https://abseil.io/tips/117

Reply:
I considered that but there are users of nontrivial members in intermediate classes, eg IndexerASTConsumer::HandleTranslationUnit. (AIUI even the trivial ones are in undefined state after a std::move; it seems better to just keep them in IndexerOptions rather than pick each of them out before the move.)

Reply:
fair enough; how would you feel about annotating with `ABSL_ATTRIBUTE_LIFETIME_BOUND`?

Reply:
SGTM